### PR TITLE
Non-linear torque

### PR DESCRIPTION
**Pull request by MoreTore**
- This fixes the jerky steering when making turns. Specifically at low speeds.
- Retune is required. This can be done by allowing torqued to converge by lifting the sanity limits.
- latAccelFactor will drop after tuning, allowing torqued to converge properly.
- A dramatic improvement in steering feel and stability.
- latAngleFactor should be added to torque_data. A zero factor means linear and anything above zero is non-linear. Then this can be added to the default torque_from_lateral_accel() function without needing the linear flag. But that's your choice to make.

### CODE REVIEW

File: selfdrive/car/interfaces.py [hunk 1]
```diff
@@ -140,9 +140,22 @@ def torque_from_lateral_accel_linear(lateral_accel_value, torque_params, lateral
...
+  def torque_from_lateral_accel_non_linear(lateral_accel_value, torque_params, lateral_accel_error, lateral_accel_deadzone, steering_angle, friction_compensation):
+    # The is a non-linear relationship between torque and lateral acceleration (accounting for road roll and steering friction)
+    # This function adjusts the lateral accel factor based on the steering angle
+    friction_interp = interp(
+      apply_center_deadzone(lateral_accel_error, lateral_accel_deadzone),
+      [-FRICTION_THRESHOLD, FRICTION_THRESHOLD],
+      [-torque_params.friction, torque_params.friction]
+    )
+    lat_factor = torque_params.latAccelFactor * ((abs(steering_angle) * torque_params.latAngleFactor) + 1) # This is the non-linear part. Adding 1 to avoid a zero factor seems apropriate
+    friction = friction_interp if friction_compensation else 0.0
```
Comment (sshane) [inline selfdrive/car/interfaces.py:154]: We're currently trying to make the Bolt good, and we think it has a non-linear response as well. Maybe something can be made generic. @nuwandavek might have some input

Comment (MoreTore) [reply selfdrive/car/interfaces.py:154]: Did you guys get around to trying this on the bolt?

File: selfdrive/controls/lib/latcontrol_torque.py [hunk 2]
```diff
@@ -64,10 +65,10 @@ def update(self, active, CS, VM, params, last_actuators, steer_limited, desired_
       error = setpoint - measurement
       gravity_adjusted_lateral_accel = desired_lateral_accel - params.roll * ACCELERATION_DUE_TO_GRAVITY
       pid_log.error = self.torque_from_lateral_accel(error, self.torque_params, error,
-                                                     lateral_accel_deadzone, friction_compensation=False)
+                                                     lateral_accel_deadzone, steering_angle, friction_compensation=False)
```
Comment (MoreTore) [inline selfdrive/controls/lib/latcontrol_torque.py:68]: Send a zero for steering angle seems to improve control.

**Comment by MoreTore** [comment]
commit [89158ae](https://github.com/commaai/openpilot/pull/27016/commits/89158ae50632010b9b5e44085d37d3539d22bc77) removes the linear calculation for pid error.

**Comment by twilsonco** [comment]
@MoreTore The process Comma is using could be applied to any car. It's missing the speed dependence but is still hypothetically general. Have you tried using their function to fit data from your car? If you're interested, I'd be happy to collect logs from you and see how your car compares to others. Some cars (like Bolt EUV) have non-linear w.r.t lat accel but no speed dependence. Other cars (like Ram 1500) have speed dependence but it's linear at each step. 

Here's my Bolt EUV fit vs comma's current one.
mph


And here's my current Ram 1500 fit.
mph


I've also found that roll compensation is speed dependent, and I've accurately fit the lateral jerk dependence for FF. In the Volt I'm using a 3-part composite lat accel + lat jerk + lat gravitational accel FF and it's been performing great, a big improvement over only having fit the lat accel component.

Meanwhile, I'm moving to a neural network approach in order to get a more robust FF that takes into account things like less roll compensation being necessary at high lat accel, and other interdependencies between what are considered the independent variables of FF. This is WIP but the preliminary results are fantastic, and importantly, the NN FF is recovering the same behavior w.r.t lateral jerk and lateral gravitational accel as that which resulted from my manual fitting processes, so it serves as a sort of validation of the behavior overall.

Here's the neural network feedforward I'm currently testing in the Volt.
voltlat

This too will be a general feedforward solution for all cars. Let me know if you want to try one for your car.

**Comment by MoreTore** [comment]
If you believe you can improve the steering control differently than what I came up with I would be very interested in testing.

Here are some drives.

164080f7933651c4|2023-04-07--07-26-08--0
164080f7933651c4|2023-04-07--07-14-42--0
164080f7933651c4|2023-02-21--12-51-09--0

**Comment by MoreTore** [comment]
> @MoreTore The process Comma is using could be applied to any car. It's missing the speed dependence but is still hypothetically general. Have you tried using their function to fit data from your car? If you're interested, I'd be happy to collect logs from you and see how your car compares to others. Some cars (like Bolt EUV) have non-linear w.r.t lat accel but no speed dependence. Other cars (like Ram 1500) have speed dependence but it's linear at each step.
> 
> Here's my Bolt EUV fit vs comma's current one. mph [ mph ](https://user-images.githubusercontent.com/7284371/230622932-b6e253a1-4c73-42d4-8ce5-9ff35cc7968c.gif) [ ](https://user-images.githubusercontent.com/7284371/230622932-b6e253a1-4c73-42d4-8ce5-9ff35cc7968c.gif)
> 
> And here's my current Ram 1500 fit. mph [ mph ](https://user-images.githubusercontent.com/7284371/230623013-ad0d76d2-d6af-4af4-a68a-6e815cfedb20.gif) [ ](https://user-images.githubusercontent.com/7284371/230623013-ad0d76d2-d6af-4af4-a68a-6e815cfedb20.gif)
> 
> I've also found that roll compensation is speed dependent, and I've accurately fit the lateral jerk dependence for FF. In the Volt I'm using a 3-part composite lat accel + lat jerk + lat gravitational accel FF and it's been performing great, a big improvement over only having fit the lat accel component.
> 
> Meanwhile, I'm moving to a neural network approach in order to get a more robust FF that takes into account things like less roll compensation being necessary at high lat accel, and other interdependencies between what are considered the independent variables of FF. This is WIP but the preliminary results are fantastic, and importantly, the NN FF is recovering the same behavior w.r.t lateral jerk and lateral gravitational accel as that which resulted from my manual fitting processes, so it serves as a sort of validation of the behavior overall.
> 
> Here's the neural network feedforward I'm currently testing in the Volt. voltlat
> 
> This too will be a general feedforward solution for all cars. Let me know if you want to try one for your car.

I see in your repo you are loading a CSV file. Where are you getting this file and what does it consist of? 
```
function main()
  
  data = load_data("/Users/haiiro/NoSync/voltlat_large.csv", false)

  model, input_mean, input_std, X_train, y_train, X_test, y_test = train_model("/Users/haiiro/NoSync/voltlat", false, data)
  
  test_plot_model(model, "/Users/haiiro/NoSync/voltlat", X_train, y_train, X_test, y_test, input_mean, input_std)
  
end
```

**Comment by twilsonco** [comment]
The proven process I use takes about 20+ hours of logs to adequately sample the problem space. I have log acquisition scripts for collecting logs over ssh or comma connect from dozens of users as I apply this to more makes. Or you can run the scripts and upload logs to me. 

The code you pasted is from the Julia repo I’m using to do the neural network fits, unrelated to the proven method. But that csv contains steer torque, lateral acceleration, lateral jerk, longitudinal speed and acceleration, and lateral gravitational acceleration, used to train the NN model for making a function for steer command as a function of the rest.

**Comment by cbenhagen** [comment]
@MoreTore why was this closed?

**Comment by MoreTore** [comment]
> @MoreTore why was this closed?

Honestly, I screwed up and force-pushed to this branch when updating to the 9.2 torque controller. I meant to open a separate PR that was more incremental than this one and bump it up to the top of the PR list.

**Comment by rockerrock137** [comment]
@MoreTore does this look like it could be solved by your work? (Ford research vehicle)
image

**Comment by MoreTore** [comment]
> @MoreTore @twilsonco  does this look like it could be solved by your work? (Ford research vehicle) I would be happy to share logs and a pizza for your time

If there is real pizza involved I would be glad to take a look. There isn't enough information here  for me to make a determination. It does look similar to what I have seen in my car. If you could share a drive I could look. It would also save me some time if you could post a YT video of the issue when it happens.

**Comment by rockerrock137** [comment]
> > @MoreTore @twilsonco  does this look like it could be solved by your work? (Ford research vehicle) I would be happy to share logs and a pizza for your time
> 
> If there is real pizza involved I would be glad to take a look. There isn't enough information here for me to make a determination. However, it does look similar to what I have seen in my car. If you could share a drive I could look. It would also save me some time if you could post a YT video of the issue when it happens.

Allow me to collect some data, my discord username is rockerrock137

I am going to try out the non-linear function from the bolt now:
https://github.com/rockerrock137/openpilot/commit/9432f4bba28612539661789421681dcf17d8f36e

Also going to try and "lift the sanity values" but I'm unsure what direction and how much I should change values, will try:
https://github.com/rockerrock137/openpilot/commit/600fb676a1c79f8646eefb5029e440ddf4d49045

**Comment by MoreTore** [comment]
> Also going to try and "lift the sanity values" but I'm unsure what direction and how much I should change values, will try: [rockerrock137@600fb67](https://github.com/rockerrock137/openpilot/commit/600fb676a1c79f8646eefb5029e440ddf4d49045)

You should just remove the sanity checks for now and watch the values in plotjuggler.

**Pull request closed without merging**